### DISCUSSION TITLE
Differences in results compared to Lucene

### DISCUSSION THREAD

**Discussion by ignorejjj** [xhluca/bm25s#53] [Q&A] [answered]
This is a very good library, very lightweight and easy to install. I want to use it in my project to replace the use of Pyserini (based on Lucene and Java). 

I tested the results of BM25 under the same corpus and found some differences in the results (use your example code for nq). Overall, the top 10 result documents are similar, but there are some differences in order. Is this normal?

**Comment by xhluca** [comment, marked as answer]
Yeah that's normal, we can see the difference in NDCG@10 in the report: https://arxiv.org/abs/2407.03618

There could be a few reasons:
1. We use a different tokenizer from the Lucene library, which I was not able to find the exact implementation for
2. The use of stemmer and choice of stopwords can affect the final scores
3. Our scoring method might differ, we base on [the Kamphuis+ survey](https://cs.uwaterloo.ca/~jimmylin/publications/Kamphuis_etal_ECIR2020_preprint.pdf) whereas Pyserini uses Lucene behind the scene, which changes its algorithm (if you can find it, please feel free to share it here)

Overall I think the best way to verify if this is the exact BM25 scoring you want is to implement bm25 manually (should be less than 50lines) and compare it against one of the supported variant here. You can easily add your variant here: https://github.com/xhluca/bm25s/blob/0a49c62ce4d98869ddb3f82a6be2294fc09c7611/bm25s/scoring.py#L99-L160

**Reply by ignorejjj** [reply]
Thank you for your reply! I will conduct some experiments later.
